### PR TITLE
Improvement/issues-29-31

### PR DESCRIPTION
**Pull request by Blu3spirits**

### CODE REVIEW

File: Dockerfile
Comment (Kiro47) [inline Dockerfile]: This kind of ruins the entire point of keeping a `package-lock.json`.

**Comment by Blu3spirits** [comment]
May want to wait on this, until after @Kiro47 merges their PR #34 since it will change the csv file dropped.

File: Dockerfile
Comment (Kiro47) [inline Dockerfile]: :slightly_smiling_face:

File: Dockerfile
Comment (Kiro47) [inline Dockerfile]: It'll probably be better to form this into a multi-stage container so that way the final image is as light as possible with minimal things installed.  It'll also prevent us from committing the heresy that is curl piping into a shell.

The following is an example of how I've done it with Go before, we can also just use a Node14 container due to the building.
https://gist.github.com/Kiro47/f176ca085c9a40fff3be54ad4a07ed77

File: course_gather/settings/prod_settings.py
Comment (Kiro47) [inline course_gather/settings/prod_settings.py]: A note for this should end up being added to the `README.md`

**Review by Kiro47 — CHANGES_REQUESTED** [review]

**Comment by Kiro47** [comment]
> May want to wait on this, until after @Kiro47 merges their PR #34 since it will change the csv file dropped.

I can always just make this change on my branch if you end merging before I do, it's a small aspect.

File: Dockerfile
Comment (codetheweb) [inline Dockerfile]: I don't know if this belongs here; it might make more sense to do the initial / scheduled scrapes after the image is created with `docker exec`.

File: Dockerfile
Comment (codetheweb) [inline Dockerfile]: We should probably explicitly create a volume for the DB.

File: Dockerfile
Comment (codetheweb) [inline Dockerfile]: Is there a reason we're using rsync here?

I think we could just do `cp -r build/* /app/static` and then `mv /app/static/static/* /app/static` (and optionally `rm -r /app/static/static`).

**Review by codetheweb — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: rsync was the first tool I thought of with exclusion capabilities that weren't annoying to use.

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: For clarification's sake, do you mean a Docker volume?

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: In a raw docker environment, I would agree. However more the point of this was a "Get me an environment that will work in one shot". Rather than:
Build the image
Deploy the image
Setup your own custom docker execs that go into the image and modify the data state.

File: Dockerfile
Comment (codetheweb) [reply Dockerfile]: Yeah, sorry. Just so we can have a persistent DB between image changes.

File: Dockerfile
Comment (codetheweb) [reply Dockerfile]: I gotcha.  As long as something like `docker exec courses python3 ./banwebScrape.py && python3 ./manage.py custom_db_sync /tmp/writer.csv` works then that's fine to have seed data in the image.

(Maybe have an example scrape / DB update command in the README?)

File: Dockerfile
Comment (Blu3spirits) [inline Dockerfile]: Thinking more about this, I'm going to generalize this into just `SECRET_KEY`

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: https://github.com/Kiro47/MTU-Transfer-Course-Gatherer#run-the-scrape-script-to-collect-the-up-to-date-course-list

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: I think for the sake of expansion I'd like to move this to another issue. Ideally, I'd like to use a real DBMS instead of SQLITE, and have that hosted in another container. (Maybe the same container for dev?)

Comment (codetheweb) [reply Dockerfile]: 👍 splitting stuff into Dockerfile.dev and Dockerfile works for me.

Maybe for the prod Dockerfile we could [omit the seed data](https://github.com/Kiro47/MTU-Transfer-Course-Gatherer/pull/33#discussion_r419166234).

So Dockerfile.dev would be batteries-included quickstart, Dockerfile would be recommended for prod envs.

Comment (Blu3spirits) [reply Dockerfile]: Created Issue #35

File: Dockerfile
Comment (codetheweb) [inline Dockerfile]: Should we be cloning here, or can we just do `COPY`?

Cloning won't easily allow us to test out different branches.

Comment (Blu3spirits) [reply Dockerfile]: I think yes. (Though it's set to the wrong repo)
And if you're downloading this Dockerfile, you're not running the dev image, you're going to be running the prod image. Therefore you shouldn't be editing anything but the environment variables

Comment (Blu3spirits) [reply Dockerfile]: Thinking more on this. Maybe not. Going to just use COPY and we'll see how clunky, if at all, it ends up being.

Comment (codetheweb) [reply Dockerfile]: If we do use COPY, we should copy package* files in a separate command, like [described here](https://nodejs.org/de/docs/guides/nodejs-docker-webapp/#creating-a-dockerfile).  Speeds up subsequent builds (assuming build layers are cached and dependencies haven't changed).

Also probably want to add a `.dockerignore` for `node_modules`.

File: Dockerfile
Comment (codetheweb) [inline Dockerfile]: https://github.com/Kiro47/MTU-Transfer-Course-Gatherer/pull/33#discussion_r419158929

Comment (Blu3spirits) [reply Dockerfile]: :facepalm:

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: @codetheweb Broke it up even than that.

File: .dockerignore
Comment (codetheweb) [inline .dockerignore]: Just tested it out, needs `*/node_modules` to be properly ignored.

File: Dockerfile
Comment (codetheweb) [reply Dockerfile]: Breaking up `COPY` commands doesn't really do much if there's nothing between them 😛 

I was thinking more like 
```dockerfile
WORKDIR /app/webapp/

COPY webapp/package*.json .

RUN npm install --only=prod

COPY webapp/* .

# Build up the frontend
ARG REACT_APP_ENDPOINT
ENV REACT_APP_ENDPOINT=$REACT_APP_ENDPOINT

RUN npm run build
```

File: Dockerfile
Comment (codetheweb) [inline Dockerfile]: We should only be copying the build output, this pulls in the entire `node_modules` folder.

File: .dockerignore
Comment (Blu3spirits) [reply .dockerignore]: ecf6404

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: ecf6404 
Only copying static over now

**Comment by Blu3spirits** [comment]
Slimmed-down quite a lot with your guys' help!

```
course_gather                              latest              3a99e7ab9378        3 minutes ago       992MB
course_gather_dev                          latest              0ad64e380fc6        3 days ago          1.46GB
```
Course gather dev is before all of these optimizations.

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: There are things between them?
Doing that your way would include things such as the README.md file that's under webapp/

Which then gets discarded later by only grabbing `/app/static/` from builder. Why even use the glob if all it does is result in discarded files?

File: Dockerfile
Comment (Kiro47) [inline Dockerfile:55]: `/app/writer.csv` Seems like the wrong path here, shouldn't it still be `/tmp/writer.csv` atm?

File: Dockerfile
Comment (Kiro47) [inline Dockerfile]: Doesn't look like we're using this in the build after this.

File: README.md
Comment (Kiro47) [inline README.md:74]: What should `SECRET_KEY` be? :thinking:

File: Dockerfile
Comment (codetheweb) [reply Dockerfile]: 👍

File: Dockerfile
Comment (codetheweb) [reply Dockerfile]: Yeah, you can change that last COPY if you want.

What I was trying to get at is that with cached build layers, a change in any of the COPYed files triggers all the commands below it to be re-run, instead of pulling from cache.  This means that if you copy the source code at the same time you copy `package.json`, a change in the source code will trigger `npm install` again (even though dependencies haven't changed).

http://bitjudo.com/blog/2014/03/13/building-efficient-dockerfiles-node-dot-js/

File: README.md
Comment (Blu3spirits) [reply README.md:74]: Literally anything within those quotes to make a string. It could be the national anthem for all Django cares.

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: Good point, should we add tests to the Docker flow? I don't see value in it currently, but if we did React tests I could see value.

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile:55]: Oddly enough, this worked, and /tmp/writer.csv didn't. :confused:

File: Dockerfile
Comment (codetheweb) [reply Dockerfile]: Personally, I would say that code testing shouldn't be part of the Dockerfile. Tests should be run before building the container. But that doesn't always happen so 🤷

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: Took me a bit of playing to understand what you were getting at. Think I finally got it though.

Comment (codetheweb) [reply Dockerfile]: Looks much better 👍.

The JSX comment just needs to be updated.
And I think line 9 can go before 7.

File: Dockerfile
Comment (codetheweb) [reply Dockerfile:55]: The `/tmp` directory in Docker doesn't really [work the way you'd expect](https://stackoverflow.com/a/55104489/12638523).

File: Dockerfile
Comment (Blu3spirits) [reply Dockerfile]: I would disagree since the flow is:
1. Copy from source
2. CD into source
3. NPM install

**Comment by Blu3spirits** [comment]
I think github actions may have shit the bed again tonight. Going to ignore that failed run for now because I cannot reproduce that error on my local branch.

File: Dockerfile
Comment (codetheweb) [reply Dockerfile]: 🤷 alright, doesn't really matter.

**Comment by codetheweb** [comment]
Yeah, I dunno what's going on. Tests pass locally. Same issue in https://github.com/Kiro47/MTU-Transfer-Course-Gatherer/pull/39.

**Review by codetheweb — APPROVED** [review]

**Comment by Blu3spirits** [comment]
Alright, thanks to @Kiro47 we're good now to merge this after I resolve the conflicts

**Review by Kiro47 — APPROVED** [review]

**Pull request merged**
